src/compiler/crystal/semantic/ast.cr

```diff
@@ -13,6 +13,10 @@ module Crystal
       ::raise exception_type.for_node(self, message, inner)
     end
 
+    def unreachable!(message = "unreachable", inner = nil, exception_type = Crystal::TypeException)
+      ::raise exception_type.for_node(self, "BUG: #{message}", inner)
```

Comment:
Shouldn't this raise an `UnreachableError` to be compatible with `::unreachable!`? Otherwise it gets confusing.